R/print.data.table.R

```diff
@@ -64,15 +65,27 @@ print.data.table = function(x, topn=getOption("datatable.print.topn"),
     }
     return(invisible(x))
   }
+  index_dt = NULL
+  if (show.indices) {
+    if (is.null(indices(x))) {
+      show.indices = FALSE
+    } else {
+      index_dt <- as.data.table(attributes(attr(x, 'index')))
+      print_names <- paste0("index", if (ncol(index_dt) > 1L) seq_len(ncol(index_dt)) else "", ":", sub("^__", "", names(index_dt)))
+      setnames(index_dt, print_names)
+    }
+  }
   n_x = nrow(x)
   if ((topn*2L+1L)<n_x && (n_x>nrows || !topnmiss)) {
     toprint = rbindlist(list(head(x, topn), tail(x, topn)), use.names=FALSE)  # no need to match names because head and tail of same x, and #3306
     rn = c(seq_len(topn), seq.int(to=n_x, length.out=topn))
     printdots = TRUE
+    if (show.indices) toprint = cbind(toprint, rbindlist(list(head(index_dt, topn), tail(index_dt, topn)), use.names=FALSE))
```

Comment:
two suggested alternatives to maybe make this a bit cleaner:

#### 1

```r
if (show.indices) index_dt = rbindlist(...)

if (show.indices) toprint = cbind(toprint, index_dt) # i.e. use the same expression just once, outside the if/else branch
```

#### 2

```r
idx = c(seq_len(topn), seq(to=nrow(x), length.out=topn))
toprint = x[idx, ]
if (show.indices) toprint = cbind(toprint, index_dt[idx, ])
```

---

At a glance the latter is a bit more readable and also more efficient than the `rbindlist()` approach:

```r
DT=setDT(lapply(1:10, \(.) rnorm(1e7)))
microbenchmark(times = 20,
  rbindlist=rbindlist(list(head(DT, 5), tail(DT, 5)), use.names=FALSE),
  seq={idx = c(seq_len(5), seq(nrow(DT)-5+1, length.out=5)); DT[idx, ]})
# Unit: microseconds
#       expr     min      lq     mean  median       uq     max neval cld
#  rbindlist 489.090 518.006 574.3189 537.609 589.5205 991.939    20   b
#        seq 171.324 188.492 215.6645 206.098 233.0890 314.695    20  a 
```

Reply:
the latter works for me!